firmware/application/apps/ble_app.cpp

```diff
@@ -0,0 +1,403 @@
+ * the Free Software Foundation, Inc., 51 Franklin Street,
+ * Boston, MA 02110-1301, USA.
+ */
+
+#include "ble_app.hpp"
+#include "ui_modemsetup.hpp"
+
+#include "modems.hpp"
+#include "audio.hpp"
+#include "rtc_time.hpp"
+#include "baseband_api.hpp"
+#include "string_format.hpp"
+#include "portapack_persistent_memory.hpp"
+
+#define USE_CONSOLE 0
+
+using namespace portapack;
+using namespace modems;
+
+void BLELogger::log_raw_data(const std::string& data) {
+    log_file.write_entry(data);
+}
+
+std::string pad_string_with_spaces(int snakes) {
+    std::string paddedStr(snakes, ' ');
+    return paddedStr;
+}
+
+namespace ui {
+template <>
+void RecentEntriesTable<BleRecentEntries>::draw(
+    const Entry& entry,
+    const Rect& target_rect,
+    Painter& painter,
+    const Style& style) {
+    std::string line = to_string_hex(entry.macAddress & 0xFF, 2);
+
+    line += ":" + to_string_hex((entry.macAddress >> 8) & 0xFF, 2);
+    line += ":" + to_string_hex((entry.macAddress >> 16) & 0xFF, 2);
+    line += ":" + to_string_hex((entry.macAddress >> 24) & 0xFF, 2);
+    line += ":" + to_string_hex((entry.macAddress >> 32) & 0xFF, 2);
+    line += ":" + to_string_hex((entry.macAddress >> 40), 2);
+
+    // Handle spacing for negative sign.
+    uint8_t db_spacing = entry.dbValue > 0 ? 7 : 6;
+
+    // Pushing single digit values down right justified.
+    if (entry.dbValue > 9 || entry.dbValue < -9) {
+        db_spacing--;
+    }
+
+    line += pad_string_with_spaces(db_spacing) + to_string_dec_int(entry.dbValue);
+
+    line.resize(target_rect.width() / 8, ' ');
+    painter.draw_string(target_rect.location(), style, line);
+}
+
+BleRecentEntryDetailView::BleRecentEntryDetailView(NavigationView& nav, const BleRecentEntry& entry)
+    : nav_{nav},
+      entry_{entry} {
+    add_children({&button_done,
+                  &labels});
+
+    button_done.on_select = [this](const ui::Button&) {
+        nav_.pop();
+    };
+}
+
+void BleRecentEntryDetailView::update_data() {
+}
+
+void BleRecentEntryDetailView::focus() {
+    button_done.focus();
+}
+
+Rect BleRecentEntryDetailView::draw_field(
+    Painter& painter,
+    const Rect& draw_rect,
+    const Style& style,
+    const std::string& label,
+    const std::string& value) {
+    const int label_length_max = 4;
+
+    painter.draw_string(Point{draw_rect.left(), draw_rect.top()}, style, label);
+    painter.draw_string(Point{draw_rect.left() + (label_length_max + 1) * 8, draw_rect.top()}, style, value);
+
+    return {draw_rect.left(), draw_rect.top() + draw_rect.height(), draw_rect.width(), draw_rect.height()};
+}
+
+void BleRecentEntryDetailView::paint(Painter& painter) {
+    View::paint(painter);
+
+    const auto s = style();
+    const auto rect = screen_rect();
+
+    auto field_rect = Rect{rect.left(), rect.top() + 16, rect.width(), 16};
+
+    uint8_t type[5];
+    uint8_t length[5];
+    uint8_t data[5][40];
+
+    int currentByte = 0;
```

Comment:
nit: you don't _have_ to pack all your decls at the top of the scope in C++.
It's not something I'd mandate either way, but I typically keep the decl as close to where it's used as possible so it's near its context.